### PR TITLE
feat(split): parent group update partition count

### PR DESCRIPTION
**Pull request by hycdong**
## Simple partition split process
1. meta receives client partition split request, and change partition count #286
2. replica notices partition count changed during on_config_sync #653 
3. parent partition create child partition #291
4. parent prepare states for child to learn #299
5. child partition async learn states from parent #309 #319
6. child notify parent catch up #390
7. update child group partition count #645 
8. register child partition #391
9. **update parent group partition count**

More partition split discussion in issue #69 and partition split design doc
This pr solves the part of ninth step of partition split, which is bold in process description.

## What this pr solve
When primary parent registers child succeed, it will update its own partition_count, then broadcast this to secondary through group_check. Besides, this pull request also adds related unit test.

### CODE REVIEW

File: src/replica/split/replica_split_manager.cpp [hunk 1]
```diff
@@ -1101,5 +1103,20 @@ void replica_split_manager::child_handle_async_learn_error() // on child partiti
...
 
+// ThreadPool: THREAD_POOL_REPLICATION
+void replica_split_manager::secondary_parent_handle_split(
+    const group_check_request &request,
+    /*out*/ group_check_response &response) // on secondary parent partition
+{
+    if (request.app.partition_count ==
+        _replica->_app_info.partition_count * 2) { // secondary update partition count
+        update_local_partition_count(request.app.partition_count);
+        parent_cleanup_split_context();
```
Comment (neverchanje) [inline src/replica/split/replica_split_manager.cpp:1114]: Why we cleanup split context? The secondary split is proceeding. I guess `parent_cleanup_split_context` means initialization before split?

Comment (hycdong) [reply src/replica/split/replica_split_manager.cpp:1114]: `parent_cleanup_split_context` will be called when parent start to split and finish split.

File: src/replica/replica_check.cpp [hunk 2]
```diff
@@ -175,6 +176,7 @@ void replica::on_group_check(const group_check_request &request,
         if (request.last_committed_decree > last_committed_decree()) {
             _prepare_list->commit(request.last_committed_decree, COMMIT_TO_DECREE_HARD);
         }
+        _split_mgr->secondary_parent_handle_split(request, response);
```
[the remark this answers is not in the document]
Comment (hycdong) [reply src/replica/replica_check.cpp:179]: This function will be used to different split status, such as pause split, cancel split, not only including starting split.

[the remark this answers is not in the document]
Comment (neverchanje) [reply src/replica/replica_check.cpp:179]: Then you should add comments here anyway. People need to understand by any means. How can I know what this line does by simply seeing such a function name `secondary_parent_handle_split`, which indicates nothing but split, secondary parent? 

```suggestion
        // the group check may trigger starting/canceling/pausing a split on the secondary.
        _split_mgr->secondary_parent_handle_split(request, response);
```

File: src/replica/replica_check.cpp [hunk 3]
```diff
@@ -175,6 +176,8 @@ void replica::on_group_check(const group_check_request &request,
         if (request.last_committed_decree > last_committed_decree()) {
             _prepare_list->commit(request.last_committed_decree, COMMIT_TO_DECREE_HARD);
         }
+        // the group check may trigger split on the secondary.
```
Comment (neverchanje) [inline src/replica/replica_check.cpp:179]:
```suggestion
        // the group check may trigger starting/canceling/pausing a split on the secondary.
```

**Review by neverchanje — APPROVED** [review]

**Review by levy5307 — DISMISSED** [review]

**Review by acelyc111 — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]

**Review by neverchanje — APPROVED** [review]

**Pull request merged**